File: 53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python
#
# [53] Maximum Subarray
#

# @lc code=start
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        preSum = nums[0]
        maxSum = preSum
        for num in nums[1:]:
            preSum = max(num, preSum+num)
            maxSum = max(maxSum, preSum)
        return maxSum
# @lc code=end

File: 53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */
#include <vector>
using namespace std;

// @lc code=start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size()==0)
            return 0;
        int preSum = nums[0];
        int maxSum = preSum;
        for (int i = 1; i < nums.size(); i++) {
            preSum = preSum > 0 ? preSum + nums[i] : nums[i];
            maxSum = max(maxSum, preSum);
        }
        return maxSum;
    }
};
// @lc code=end

